### DISCUSSION TITLE
Custom Layersets

### DISCUSSION THREAD

**Discussion by KoniWatson** [rustprooflabs/pgosm-flex#225] [Q&A] [answered]
Hi All!

I am trying use a custom layerset and I feel like I am missing something because it is not working. 

At the moment I get the following error:

_Layerset INI path set to /custom-layerset/
Loading config: /custom-layerset/test.ini
2022-02-01 14:46:04  ERROR: Error loading lua config: /usr/local/share/lua/5.2/inifile.lua:88: attempt to index field '?' (a nil value).

END PgOSM Flex output
2022-02-01 14:46:04,739:ERROR:pgosm-flex:pgosm_flex:Failed to run osm2pgsql. Return code: 1
Failed to run osm2pgsql. Return code: 1 - Check the log output for details._

I get this when following the 'Use Custom Layerset' section of the docs -- > [https://github.com/rustprooflabs/pgosm-flex/blob/main/docs/DOCKER-RUN.md#use-custom-layersets](url)

I have attempted changing the permissions of the .ini file and folder that it is in. I have copy and pasted one of the built in layersets into a new file and tried to load that to see if there was an issue within my .ini file but got the same error.

What I am using:

- Ubuntu 21.10
- osm2pgsql 1.4.1
- PostgreSQL 14.1 
- PostGIS 3.2
- Pgosm-Flex Docker Image

If anyone is able to help that would be greatly appreciated!

Kindest Regards,
Koni

**Comment by rustprooflabs** [comment, marked as answer]
Hi - I'm sorry you are having this issue!  Can you share the contents of your ini file?  It appears the error is in parsing the contents of the ini file. I just followed the steps with a copy of `minimal.ini` without issue. Though, considering I wrote those steps it's quite possible I'm completely missing something obvious that happens to work on my machines.

**Reply by KoniWatson** [reply]
Thank you very much for getting back to me! I have attached the layerset.ini file as a txt:
[layerset.txt](https://github.com/rustprooflabs/pgosm-flex/files/7987797/layerset.txt)

**Reply by rustprooflabs** [reply]
This is a strange one. I downloaded that .txt, renamed to .ini and got the same error you reported. I tried copy/pasting an existing layerset of mine into that file and then it failed with the same message (even though the original works). But, when I copied/pasted the contents of your file into a new file, it works without issue.  The layer listing reported from the osm2pgsql run is:

```bash
PgOSM-Flex version:	0.4.3	208d31b
ensuring pgosm_flex table exists.
Including infrastructure
Including landuse
Including natural
Including place
Including poi
Including public_transport
Including road
Including road_major
Including tags
Including water
```

With all that, it does not appear to be an issue with the contents of the file, but the file itself.  I have no clue what's actually wrong with it unfortunately, though I don't believe it's related to permissions.  I'd delete that file, start fresh, and hope that it was some strange oddity.

Also, I don't know that we can even improve the error message reported when it fails to parse the INI file. The error comes from the `inifile.parse()` command [here](https://github.com/rustprooflabs/pgosm-flex/blob/main/flex-config/layerset.lua#L28).  The error reported is the only detail available, so we're left knowing it's an issue with parsing the INI file and nothing else.

**Comment by KoniWatson** [comment]
That is very strange but weirdly gives me peace of mind that it wasn't something I was doing wrong! Thank you very much for looking into it for me and will attempt what you have suggested.
